starknet-accounts/src/single_owner.rs

```diff
@@ -0,0 +1,139 @@
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum GetNonceError<P>
+where
+    P: std::error::Error + Send,
+{
+    #[error(transparent)]
+    ProviderError(P),
+    #[error("invalid response length. expected {expected} but got {actual}")]
+    InvalidResponseLength { expected: usize, actual: usize },
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ExecuteError<P, S>
+where
+    P: std::error::Error + Send,
+    S: std::error::Error + Send,
+{
+    #[error(transparent)]
+    ProviderError(P),
+    #[error(transparent)]
+    SignerError(S),
+}
+
+impl<P, S> SingleOwnerAccount<P, S>
+where
+    P: Provider + Send,
+    S: Signer + Send,
+{
+    pub fn new(provider: P, signer: S, address: UnsignedFieldElement) -> Self {
+        Self {
+            provider,
+            signer,
+            address,
+        }
+    }
+}
+
+#[async_trait]
+impl<P, S> Account for SingleOwnerAccount<P, S>
+where
+    P: Provider + Sync + Send,
+    S: Signer + Sync + Send,
+{
+    type GetNonceError = GetNonceError<P::Error>;
+    type ExecuteError = ExecuteError<P::Error, S::SignError>;
+
+    async fn get_nonce(
+        &self,
+        block_identifier: Option<BlockId>,
+    ) -> Result<UnsignedFieldElement, Self::GetNonceError> {
+        let call_result = self
+            .provider
+            .call_contract(
+                InvokeFunction {
+                    contract_address: self.address,
+                    entry_point_selector: get_selector_from_name("get_nonce").unwrap(),
+                    calldata: vec![],
+                    signature: vec![],
+                },
+                block_identifier,
+            )
+            .await
+            .map_err(GetNonceError::ProviderError)?;
+
+        if call_result.result.len() == 1 {
+            Ok(call_result.result[0])
+        } else {
+            Err(GetNonceError::InvalidResponseLength {
+                expected: 1,
+                actual: call_result.result.len(),
+            })
+        }
+    }
+
+    async fn execute(
+        &self,
+        to: UnsignedFieldElement,
+        selector: UnsignedFieldElement,
+        calldata: &[UnsignedFieldElement],
+        nonce: UnsignedFieldElement,
+    ) -> Result<AddTransactionResult, Self::ExecuteError> {
+        let message_hash = compute_hash_on_elements(&[
+            self.address,
+            to,
+            selector,
+            compute_hash_on_elements(calldata),
+            nonce,
+        ]);
+        let signature = self
+            .signer
+            .sign_hash(&message_hash)
+            .await
+            .map_err(ExecuteError::SignerError)?;
+
+        let mut execute_calldata = vec![to, selector, calldata.len().into()];
+        for item in calldata.iter() {
+            execute_calldata.push(*item);
+        }
+        execute_calldata.push(nonce);
+
+        self.provider
+            .add_transaction(
+                TransactionRequest::InvokeFunction(InvokeFunctionTransactionRequest {
```

Comment:
Outside of the scope of this PR, but the renaming of `InvokeFunctionTransaction` to `InvokeFunctionTransactionRequest` got me really confused when I was navigating the code (I was also looking at the `InvokeFunction` in call_contract.rs which has the same structure, so I was like 🤯 ). Is it common in Rust to rename types like this in `mod.rs`?

Reply:
I think it's common in Rust to name types as simple as possible _inside_ modules. For example you have `std::io::Error` not `std::io::StdIoError`. When re-exporting, clashes are inevitable due to this convention.

Sometimes it's not even about clashes, but rather that getting re-exported to an upper level means it loses the context provided by the module path, so you have to re-name. See `ethers-rs` for a lot of renaming in re-exports. For one, `Wallet` is exported as `LocalWallet`, etc.